File: Search/ariprog/ariprog.py
"""
ID: ali201
TASK: ariprog
LANG: PYTHON3
"""
fin = open('ariprog.in', 'r')
fout = open('ariprog.out', 'w')
lines = fin.readlines()
N = int(lines[0]) #length of progressions
M = int(lines[1]) #upper bound of bisquares, 0<=p,q<=M
fin.close()

bisquares = set() 

for i in range(M+1):
    for j in range(i, M+1):
        bisquares.add(i*i + j*j) #use bisquares to check membership

bisquares_list = list(bisquares)
bisquares_list.sort()

ret = []

for i in range(len(bisquares_list)):
    for j in range(i+1, len(bisquares_list)):
        diff = bisquares_list[j]-bisquares_list[i]
        exist = True
        for k in range(N):
            if not bisquares_list[i] + diff*k in bisquares:
                exist = False
                break
        if exist:
            ret.append([bisquares_list[i], diff])

ret.sort(key=lambda r: (r[1], r[0]))

string = []
for i in range(len(ret)):
    string.append(str(ret[i][0]) + ' ' + str(ret[i][1]) + '\n')
fout.writelines(string)
if len(string) == 0:
    fout.write('NONE\n')
fout.close()

File: Search/ariprog/ariprog.cpp
/*
ID: ali210
LANG: C++
TASK: ariprog
*/
#include <bits/stdc++.h>
using namespace std;

int N, M;
map<int, set<int>> toVal;
//set<int> keys;

int main()
{
	ifstream fin("ariprog.in");
	fin >> N >> M; //length progressions, bound
	int maxBi = 2*M*M;
	vector<bool> bisquares(maxBi + 1);
	for(int i = 0; i <= maxBi; i++){
		for(int j = 0; j <= maxBi; j++){
			bisquares[i*i+j*j] = true;
		}
	}
	fin.close();
	bool exist;
	for(int i = 0; i < maxBi; i++){
		if(!bisquares[i]){
			continue;
		}
		exist = true;
		for(int j = 1;  j <= (maxBi-i)/(N-1); j++){
			for(int k = 1; k < N; k++){
				if(!bisquares[i+j*k]){
					cout << "joe";
					exist = false;
					break;
				}
			}
			if(!exist){
				continue;
			}
			toVal[j].insert(i);
		}
	}

	ofstream fout("ariprog.out");

	for(auto const& x : toVal){
		//cout << x;
	}
	if(toVal.size() == 0){
		fout << "NONE\n";
	}
	fout.close();
	return 0;
}
